### PR TITLE
Multiple chains + deadlock removal

### PR DESCRIPTION
**Pull request by lukanus**

### CODE REVIEW

File: scheduler/destination/destination.go [hunk 1]
```diff
@@ -128,7 +128,7 @@ func (s *Scheme) Refresh(ctx context.Context) error {
 		for network, sub := range wns {
 			for _, w := range sub.Workers {
 				for _, ci := range w.ConnectionInfo {
-					k[NVCKey{network, ci.Version, ci.ChainID}] = (w.State == 1) // (lukanus): 1 is  online
+					k[NVCKey{Network: network, Version: ci.Version, ChainID: ci.ChainID}] = (w.State == 1) // (lukanus): 1 is  online
```
Comment (silentlight) [inline scheduler/destination/destination.go:131]: I think it will be better to create a constant for this magic number

**Review by silentlight — APPROVED** [review]
Looks good! Left one minor comment but approving

File: manager/client/missing.go [hunk 2]
```diff
@@ -11,11 +11,90 @@ import (
...
+			if len(sum.MissingTransactions) > 0 {
+				allRespTransactions[sum.Ident] = sum.MissingTransactions
+			}
+
+			if i == received {
+				break
+			}
+		}
+
+		if errs != nil {
```
Comment (pacmessica) [inline manager/client/missing.go:80]: this nil check is unnecessary

Comment (lukanus) [reply manager/client/missing.go:80]: Yeah, you're totally right

File: manager/store/postgres/block.go [hunk 3]
```diff
@@ -307,3 +307,35 @@ func (d *Driver) BlockTransactionCheck(ctx context.Context, blx structs.BlockWit
...
+
+	defer rows.Close()
+
+	hcp := heightCountPair{}
+	pairs := [][2]uint64{}
+	for rows.Next() {
+		if err := rows.Scan(&hcp.Height, &hcp.Count); err != nil {
+			return pairs, err
+		}
+		pairs = append(pairs, [2]uint64{hcp.Height, hcp.Count})
```
Comment (pacmessica) [inline manager/store/postgres/block.go:337]: makeshift go tuple? interesting

File: manager/client/missing.go [hunk 4]
```diff
@@ -37,6 +113,133 @@ func (hc *Client) CheckMissingTransactions(ctx context.Context, nv NetworkVersio
...
+	b, t := processMissingBlocksAndTransactions(blocks, txs, heightRange.StartHeight, heightRange.EndHeight, window, hc.logger)
+	hc.logger.Info("[Client] consistency check succeeded", zap.Uint64("start", heightRange.StartHeight), zap.Uint64("end", heightRange.EndHeight), zap.Duration("database", database), zap.Duration("all", time.Since(start)), zap.Int("missing_blocks_ranges", len(b)), zap.Int("missing_transactions_ranges", len(b)))
+	return b, t, err
+}
+
+// processMissingBlocksAndTransactions takes blocks and transactions and joins them seeking for missing pieces
+func processMissingBlocksAndTransactions(blocks, txs [][2]uint64, startHeight, endHeight, window uint64, logger *zap.Logger) (missingBlocks, missingTransactions [][2]uint64) {
+
+	missingBlocks = [][2]uint64{}
+	missingTransactions = [][2]uint64{}
```
Comment (pacmessica) [inline manager/client/missing.go:160]: I find these structs `[][2]uint64{}` a bit confusing. It's unclear to me what the data inside them represents at first glance. Maybe we should make it a type, eg `type HeightWithTxCount [2]uint64`?

File: manager/client/missing.go [hunk 5]
```diff
@@ -37,6 +113,133 @@ func (hc *Client) CheckMissingTransactions(ctx context.Context, nv NetworkVersio
...
+		return nil, nil, err
+	}
+
+	if len(blocks) == 0 {
+		missingBlocks = groupRanges([][2]uint64{{heightRange.StartHeight, heightRange.EndHeight}}, window)
+		hc.logger.Debug("[CLIENT] No blocks for ", zap.Uint64("start", heightRange.StartHeight), zap.Uint64("end", heightRange.EndHeight), zap.Duration("took", time.Since(start)))
+		return missingBlocks, missingTransactions, nil
+	}
+
+	txs, err := hc.storeEng.GetTransactionsHeightsWithTxCount(ctx, shared.BlockWithMeta{Version: nv.Version, Network: nv.Network, ChainID: nv.ChainID}, heightRange.StartHeight, heightRange.EndHeight)
```
Comment (pacmessica) [inline manager/client/missing.go:144]: is there a difference between  `NumTx` and `TxCount`?

Comment (pacmessica) [reply manager/client/missing.go:144]: from the method names, `GetBlocksHeightsWithNumTx` and `GetTransactionsHeightsWithTxCount`

File: manager/client/missing.go [hunk 6]
```diff
@@ -37,6 +113,133 @@ func (hc *Client) CheckMissingTransactions(ctx context.Context, nv NetworkVersio
...
+			break
+		}
+
+		// check if current transaction == block
+		tx := txs[txIndex]
+		var missed bool
+		if tx[0] > block[0] { // for non existing tx
+			if block[1] > 0 {
+				logger.Info("for non existing tx", zap.Any("tx", tx), zap.Any("block", block))
+				missingTransactions = append(missingTransactions, [2]uint64{block[0], block[0]})
```
Comment (pacmessica) [inline manager/client/missing.go:195]: I thought 2nd index in the tuple represented txcount. But block[0] represents a height, right? so you're adding {height, height}?

lol see how this is confusing to a layman :p

**Review by pacmessica — COMMENTED** [review]
The repo name changed! I remember it being more transaction specific. Are we planning on moving all the indexers here?

File: manager/store/postgres/block.go (hunk 3 above)
Comment (lukanus) [reply manager/store/postgres/block.go:337]: Just plain regular array :)

File: manager/client/missing.go (hunk 5 above)
Comment (lukanus) [reply manager/client/missing.go:144]: Numtxs is a column in block like numer of transactions. Count is a count operation on transaction set :)

File: manager/client/missing.go (hunk 6 above)
Comment (lukanus) [reply manager/client/missing.go:195]: Lol ;D its a range. In this particular case range of only one height :)

Comment (pacmessica) [reply manager/client/missing.go:195]: 🤦

**Pull request merged**